src/modules/ekf2/ekf2_main.cpp

```diff
@@ -1414,14 +1415,26 @@ void Ekf2::run()
 					lpos.dist_bottom = _param_ekf2_min_rng.get();
 				}
 
-				// update ground effect flag based on terrain estimation
-				if (lpos.dist_bottom_valid && lpos.dist_bottom < _param_ekf2_gnd_max_hgt.get()) {
-					_ekf.set_gnd_effect_flag(true);
+				if (!_had_valid_terrain) {
+					_had_valid_terrain = lpos.dist_bottom_valid;
 				}
 
-				// update ground effect flag based on land detector state
-				else if (vehicle_land_detected_updated && _param_ekf2_gnd_eff_dz.get() > 0.0f) {
-					_ekf.set_gnd_effect_flag(vehicle_land_detected.in_ground_effect);
+				if (vehicle_status.arming_state == vehicle_status_s::ARMING_STATE_ARMED) {
+					// set ground effect flag if vehicle is closer than a specified distance to the ground
+					if (lpos.dist_bottom_valid) {
+						_ekf.set_gnd_effect_flag(lpos.dist_bottom < _param_ekf2_gnd_max_hgt.get());
```

Comment:
I think the condition for setting ground effect compensation should also include a check on `_param_ekf2_gnd_eff_dz.get() > 0.0f`, otherwise you could end up in a situation where you set the flag but then is not applied in [vel_pos_fusion.cpp](https://github.com/PX4/ecl/blob/master/EKF/vel_pos_fusion.cpp#L111) because the parameter is 0, right?